src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -906,11 +941,21 @@ impl<'tcx> LocationTree {
         };
 
         let accessed_root_tag = accessed_root.map(|idx| nodes.get(idx).unwrap().tag);
-        if matches!(visit_children, ChildrenVisitMode::SkipChildrenOfAccessed) {
-            // FIXME: approximate which roots could be children of the accessed node and only skip them instead of all other trees.
-            return interp_ok(());
-        }
         for root in roots {
+            let tag = nodes.get(root).unwrap().tag;
+            // On a protector release access we skip the children of the accessed tag so that
+            // we can correctly return references from functions. However, if the tag has
+            // exposed children then some of the wildcard subtrees could also be children of
+            // the accessed node and would also need to be skipped. We can narrow down which
+            // child trees are children by comparing their root tag to the minimum exposed
+            // child of the accessed node. As the parent tag is always smaller than the child
+            // tag this means we only need to skip subtrees with a root tag larger than
+            // `min_exposed_child`.
```

Comment:
```suggestion
            // `min_exposed_child`. Once we find such a root, we can leave the loop because roots
            // are sorted by tag.
```